contracts/BookMarketplace.sol

```diff
@@ -44,15 +50,162 @@ contract BookMarketplace is ERC721URIStorage {
         bool sold
     );
 
+    event ListingUpdated(
+        uint256 indexed tokenId,
+        address seller,
+        address owner,
+        uint256 instantPrice,
+        uint256 startingPrice,
+        bool allowBid,
+        Bid[] bidList,
+        bool sold
+    );
+
+    /* modifiers */
+    
+    
+    // Modifier prevents function from being called by anyone other than the market owner
+    modifier onlyMarketOwner() {
+        require(
+            marketOwner == msg.sender,
+            "Only the marketplace owner can call this function."
+        );
+        _;
+    }
+
+    /* constructor */
+
     constructor() ERC721("BookExchange Tokens", "BOOK") {
-        owner = payable(msg.sender);
+        marketOwner = payable(msg.sender);
+    }
+
+    /* functions */
+
+    function getMarketOwnerAddress() public view returns(address) {
+        return marketOwner;
+    }
+
+    // change the price of a listing
+    function updateInstantPrice(uint256 tokenId, uint256 newPrice) public payable onlyMarketOwner {
+        Listing storage listing = idToMarketItem[tokenId];
+        listing.instantPrice = newPrice;
+        idToMarketItem[tokenId] = listing;
+    }
+
+    // get the price of a listing
+    function getInstantPrice(uint256 tokenId) public view returns (uint256) {
+        Listing storage listing = idToMarketItem[tokenId];
+        return listing.instantPrice;
+    }
+
+    // Adds bids to a bid list by tokenId
+    function addBid(uint256 tokenId, uint256 bidAmount) public payable {
+        Listing storage listing = idToMarketItem[tokenId];
+        require(
+            listing.allowBid == true,
+            "This item does not allow bids."
+        );
+        require(
+            listing.sold == false,
+            "This item has been sold."
+        );
+        require(
+            bidAmount >= listing.instantPrice,
+            "Bid must be greater than or equal to the current price."
+        );
+        if(hasUserBid(tokenId, msg.sender) == false) {
+            require(
+                msg.value == biddingPrice,
+                "Bid price must be paid on first bid: 0.001 ether."
+            );
+        }
+        else {
+            require(
+                msg.value == 0,
+                "Bid price must be 0 ether after first bid."
+            );
+        }
+        require(
+            bidAmount > getHighestBid(tokenId).bidAmount,
+            "Bid must be greater than the current highest bid."
+        );
+        require(
+            msg.sender != listing.seller,
+            "Seller cannot bid on their own listing."
+        );
+        require(
+            msg.sender != getHighestBidder(tokenId),
+            "Highest bidder cannot bid again."
+        );
+
+        
+        Bid memory newBid;
+        newBid.bidder = msg.sender;
+        newBid.bidAmount = bidAmount;
+        listing.bidList.push(newBid);
```

Comment:
Your require functions are on point! Love it. However, you forgot to transfer the entrance bid fees to the (contract) owner's address. An example how to do that (I copied from the tutorial which shows how to transfer listing fee to owner address): `payable(owner).transfer(listingPrice);`

Reply:
On another thought, `call` is recommended vs `transfer`. Check this article: [transfer vs send vs call](https://medium.com/coinmonks/solidity-transfer-vs-send-vs-call-function-64c92cfc878a).